[Bug 15440] lc-compile: Permit string literals with embedded nuls

peter-b opened the pull request:
Internally, lc-compile uses nul-terminated strings.  However, both the
compiled module file format and the LiveCode runtime allow strings to
contain nuls.

Previously, `"\u{0}"` evaluated to the empty string.  This was because
`UnescapeStringLiteral()` faithfully decoded `\u{0}` to the byte
string `{0x00, 0x00}`, which interpreted as a nul-terminated string is
empty.

When outputting a module file, compiler internal strings are converted
to libfoundation `MCString`.  So the nul-terminated strings used
internally by the compiler are never exposed (apart possibly via error
or diagnostic messages).

A trivial fix is therefore to use Modified UTF-8 encoding for
compiler-internal strings, where U+0000 is encoded as `{0xc0, 0x80}`.

peter-b committed:
[Bug 15440] lc-compile: Permit string literals with embedded nuls

Internally, lc-compile uses nul-terminated strings.  However, both the
compiled module file format and the LiveCode runtime allow strings to
contain nuls.

Previously, `"\u{0}"` evaluated to the empty string.  This was because
`UnescapeStringLiteral()` faithfully decoded `\u{0}` to the byte
string `{0x00, 0x00}`, which interpreted as a nul-terminated string is
empty.

When outputting a module file, compiler internal strings are converted
to libfoundation `MCString`.  So the nul-terminated strings used
internally by the compiler are never exposed (apart possibly via error
or diagnostic messages).

A trivial fix is therefore to use Modified UTF-8 encoding for
compiler-internal strings, where U+0000 is encoded as `{0xc0, 0x80}`.

runrevmark commented:
@peter-b: Just a wee thing - this works because our UTF8 decoder in libfoundation allows overlong sequences... Is this something we need to note in the MCString API which lc-compile uses?

peter-b commented:
@runrevmark I opened [bug 13892](http://quality.livecode.com/show_bug.cgi?id=13892) about that more than a year ago. :wink: As I understand it, though, overlong sequences are valid UTF-8 for most decoders' definitions of "valid".  Where did you want to note it?

runrevmark commented:
@peter-b: Maybe just add a Note to the API that lc-compile uses to build its MCStrings from UTF-8 - just so we don't 'break' it accidentally in the future if we make things more strict.

peter-b commented:
@runrevmark Sure thing.

peter-b committed:
[Bug 15440] libfoundation: Add note re overlong encodings.

Add a note to the definition of `kMCStringEncodingUTF8` pointing out
that libfoundation is currently happy to silently consume UTF8
overlong sequences, but never generates them.

peter-b commented:
@runrevmark I've added a note as suggested.

runrevmark commented:
@livecode-vulcan review ok dfb4783

livecode-vulcan commented:
💙 review by @runrevmark ok dfb4783624f6db2140374878f877bd05e5b067b8

livecode-vulcan commented:
😎 test success dfb4783624f6db2140374878f877bd05e5b067b8
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1152)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1163)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1149)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1155)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1167)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1169)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1171)
